PowerEditor/src/WinControls/Grid/ShortcutMapper.cpp

```diff
@@ -124,123 +122,176 @@ void ShortcutMapper::initBabyGrid() {
 	_babygrid.setHighlightColorProtectNoFocus(RGB(230,194,190));
 }
 
+generic_string ShortcutMapper::getTextFromCombo(HWND hCombo)
+{
+	TCHAR str[128];
+	::SendMessage(hCombo, WM_GETTEXT, 127, reinterpret_cast<LPARAM>(str));

```

Comment:
There's no point in asking for only 127 characters here. If you don't trust the Windows API, you must initialize `str[127]` to the null character before calling `SendMessage`. The `WM_GETTEXT` [documentation](https://msdn.microsoft.com/de-de/library/windows/desktop/ms632627) doesn't explicitly mention whether it copies the null character, so a super-safe bet would be to fill the complete buffer with zeroes before calling `SendMessage`.

Regarding the abstraction level: do you really need to work with `WM_GETTEXT` here? There should be a generic function that just takes a window handle and maybe a maximum string size and returns a string. This function would be useful in lots of other places and would keep this code really nice and short.

Reply:
> generic function

i.e. control APIs in `windowsx.h`.

Reply:
a lot's a code use this syntax in NPP. I keep it.